### DISCUSSION TITLE
NamedCSVReader add header line number

### DISCUSSION THREAD

**Discussion by sbi1** [osiegmar/FastCSV#66] [Q&A] [answered]
Hi, I tested the NamedCSVReader. It works fine if the header is on line 1 of the CSV file. But if the header is on line 2 or any other line number than 1 it does not work. Is it possible to define a starting position e.g. header starts on line 3?

Thank you

**Comment by osiegmar** [comment, marked as answer]
As defined in [section 2.3 of RFC 4180](https://datatracker.ietf.org/doc/html/rfc4180#section-2):

> There maybe an optional header line appearing as the *first line*

In [section 2.1.3 of current DRAFT RFC 4180-bis](https://datatracker.ietf.org/doc/html/draft-shafranovich-rfc4180-bis#section-2.1) this is updated:

> The *first record* in the file MAY be an optional header

followed by section 2.1.8:

> [...] A comment line MAY appear in any line of the file (before or after an OPTIONAL header)

This is what the NamedCsvReader supports:

```java
String csv = "# commented line\nheader\nbody\n";

NamedCsvReader.builder()
    .skipComments(true)
    .build(csv)
    .forEach(row -> System.out.println(row.getField("header")));
```

In order to skip non-RFC-compliant CSV data at the beginning you could use a BufferedReader:

```java
BufferedReader br = new BufferedReader(new StringReader(csv));

// skip arbitrary (non-CSV) data until line break
br.readLine();

NamedCsvReader.builder()
    .build(br)
    .forEach(row -> System.out.println(row.getField("header")));
```

**Comment by osiegmar** [comment]
Using FastCSV 3 (soon to be released) this would work:

```java
String csvData = """
    some arbitrary data here
    ...some other here
    header A,header B
    field A,field B
    """;

CsvReader csvReader = CsvReader.builder().build(csvData);

// Read data until you found the actual header
List<String> headerToUse = null;
for (CsvRecord csvRecord : csvReader) {
    if (csvRecord.getField(0).equals("header A")) {
        headerToUse = csvRecord.getFields();
        break;
    }
}

if (headerToUse == null) {
    throw new IllegalStateException("No header found");
}

// Initialize the NamedCsvReader with the header to use
NamedCsvReader namedCsvReader = NamedCsvReader.from(csvReader, headerToUse);

// Use the header / NamedCsvReader
for (NamedCsvRecord namedCsvRecord : namedCsvReader) {
    System.out.println(namedCsvRecord.getField("header A"));
}
```
